drivers/gpu/drm/i915/i915_active.c

```diff
@@ -130,6 +131,55 @@ static inline void debug_active_assert(struct i915_active *ref) { }
 
 #endif
 
+#if defined(__FreeBSD__)
+static struct llist_head linux_kmem_cache_list = LLIST_HEAD_INIT();
+
+struct cache_addr {
+	struct linux_kmem_cache *cache;
+	struct llist_node *next;
+};
+
+static void
+linux_kmem_cache_free_async_fn(struct irq_work *irqw)
+{
+	struct llist_node *freed_addr;
+
+	while ((freed_addr = llist_del_first(&linux_kmem_cache_list)) != NULL) {
+		struct cache_addr *caddr = (struct cache_addr *) freed_addr;
+		kmem_cache_free(caddr->cache, freed_addr);
+	}
+}
+static DEFINE_IRQ_WORK(linux_kmem_cache_free_async_work,
+	linux_kmem_cache_free_async_fn);
+
+static void
+linux_kmem_cache_free_async(struct linux_kmem_cache *c, void *m)
+{
+	struct cache_addr *ca = (struct cache_addr *)m;
+
+	if (m == NULL)
+		return;
+
+	ca->cache = c;
+
+	llist_add((void *)m, &linux_kmem_cache_list);
+	irq_work_queue(&linux_kmem_cache_free_async_work);
+}
+
+/*
+ * Critical section-friendly version of kmem_cache_free().
+ * Requires knowledge of the allocation size at build time.
+ */
+#define	KMEM_CACHE_FREE(cache, addr)	do {				\
```

Comment:
KMEM_CACHE_FREE() differs from kmem_cache_free() as first does not handle RCU cases. It would be good to leave KASSERT/BUG macro preventing passing of RCU-related caches.

Reply:
I have addressed this in my latest commit, adding a `KASSERT`. If it's wrong, comment back,